io.openems.edge.controller.ess.cycle/src/io/openems/edge/controller/ess/cycle/statemachine/Context.java

```diff
@@ -1,74 +1,103 @@
+		this.allowedDischargePower = allowedDischargePower;
+		this.parsedStartTime = parsedStartTime;
 		this.previousState = previousState;
-		this.lastStateChange = lastStateChange;
 	}
 
 	/**
-	 * Helper for a state change. If awaiting hysteresis time exceeded switches from
-	 * currentState to NextState.
-	 *
-	 * @param currentState Used to output better log.
-	 * @param nextState    state which will be switched to.
-	 * @return boolean true if change state is allowed
+	 * Helper for a state change.
+	 * 
+	 * @param nextState which will be switched to.
+	 * @return {@link State#WAIT_FOR_STATE_CHANGE}.
 	 */
-	protected boolean waitForChangeState(State currentState, State nextState) {
-		if (LocalDateTime.now(this.componentManager.getClock()).minus(Duration.ofMinutes(this.config.standbyTime()))
-				.isAfter(this.lastStateChange)) {
-			return true;
-		}
-		this.logInfo(this.log, "Awaiting hysteresis for changing from [" + currentState + "] to [" + nextState + "]");
-		return false;
+	protected State changeToNextState(State nextState) {
+		this.getParent().setNextState(nextState);
+		return State.WAIT_FOR_STATE_CHANGE;
 	}
 
-	protected int getChargePower() {
-		// get max charge/discharge power
-		var power = Math.max(this.maxChargePower, this.config.power() * -1);
-		if (this.ess instanceof HybridEss) {
-			// modify target power by DC production for HybridEss
-			var ess = (HybridEss) this.ess;
-			power -= ess.getActivePower().orElse(0) - ess.getDcDischargePower().orElse(0);
+	/**
+	 * Gets the required AC power set-point for AC- or Hybrid-ESS.
+	 * 
+	 * @param ess           the {@link ManagedSymmetricEss}; checked for
+	 *                      {@link HybridEss}
+	 * @param hybridEssMode the {@link HybridEssMode}
+	 * @param power         the configured target power
+	 * @return the AC power set-point
+	 */
+	protected Integer getAcPower(ManagedSymmetricEss ess, HybridEssMode hybridEssMode, int power) {
+		return switch (this.config.hybridEssMode()) {
+		case TARGET_AC -> this.config.power();
+		case TARGET_DC -> {
+			if (ess instanceof HybridEss) {
+				var pv = ess.getActivePower().orElse(0) - ((HybridEss) ess).getDcDischargePower().orElse(0);
+				yield pv + this.config.power();
+			}
+			yield this.config.power();
 		}
-		return power;
+		};
 	}
 
-	protected int getDischargePower() {
-		// TODO consider DC power for HybridEss
-		return Math.min(this.maxDischargePower, this.config.power());
+	/**
+	 * Helper to activate the full cycle activated or not.
+	 * 
+	 * @return true if the Controller should be executed now
+	 * @throws OpenemsNamedException on error.
+	 */
+	public boolean initializeTime() {
+		var now = LocalDateTime.now(this.componentManager.getClock());
+		var afterNow = now.isAfter(this.parsedStartTime.minusSeconds(1));
+		var beforeNow = now.isBefore(this.parsedStartTime.plusSeconds(59));
+		if (afterNow && beforeNow) {
+			return true;
+		}
+		return false;
 	}
 
+	/**
+	 * Is ESS SoC defined ?.
```

Comment:
Is this javadoc finished?
I prefer that these descriptions not end with a '?.'. You could rephrase it like 'Checks if the ESS SoC is defined.'

Also do you need list explanation and remeber the TODO

Reply:
not finished, this was a idea to think about it but this is somethng to apply on all controllers. this is not the right the place to implement it, we should do somethng on abstract class level. We can delete this TODO from here.